File: site/ros/curiosity-rover.js
// generator of correct traces for curiosity-rover.rml: speed control of Curiosity rover

'use strict';

function getRandom(min,max) {
    return min+Math.random()*(max-min);
}

function command(com) {
    event_counter++;
    //console.log(com);
}

var trace_size = Number(process.argv[2]) || 100;
var event_counter = 0;

while(event_counter<=trace_size){
    command({topic:'wheels_control',direction:'left',speed:10})
    command({topic:'wheels_control',direction:'right',speed:10})
    command({topic:'wheels_control',direction:'forward',speed:10})
    command({topic:'wheels_control',direction:'backward',speed:10})
}

File: site/ros/curiosity-rover.pl
:- module(spec, [(trace_expression/2), (match/2)]).
:- use_module(monitor(deep_subdict)).
:- use_module(library(clpr)).
match(_event, left_speed_et) :- deep_subdict(_event, _{event:"func_pre", name:"command", args:[_{topic:"wheels_control", direction:"left", speed:Val}]}),
	{(Val =< 10)}.
match(_event, right_speed_et) :- deep_subdict(_event, _{event:"func_pre", name:"command", args:[_{topic:"wheels_control", direction:"right", speed:Val}]}),
	{(Val =< 10)}.
match(_event, forward_speed_et) :- deep_subdict(_event, _{event:"func_pre", name:"command", args:[_{topic:"wheels_control", direction:"forward", speed:Val}]}),
	{(Val =< 15)}.
match(_event, backward_speed_et) :- deep_subdict(_event, _{event:"func_pre", name:"command", args:[_{topic:"wheels_control", direction:"backward", speed:Val}]}),
	{(Val =< 15)}.
match(_event, relevant_et) :- deep_subdict(_event, _{event:"func_pre", name:"command"}).
match(_event, any_et) :- deep_subdict(_event, _{}).
match(_event, none_et) :- not(match(_event, any_et)).
trace_expression('Main', Main) :- (Main=((relevant_et>>star((((left_speed_et\/right_speed_et)\/forward_speed_et)\/backward_speed_et)));1)).
